staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go

```diff
@@ -0,0 +1,336 @@
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"reflect"
+
+	"k8s.io/apiserver/pkg/cel/common"
+	celopenapi "k8s.io/apiserver/pkg/cel/openapi"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+	"k8s.io/kube-openapi/pkg/validation/strfmt"
+	"k8s.io/kube-openapi/pkg/validation/validate"
+)
+
+// schemaArgs are the arguments to constructor for OpenAPI schema validator,
+// NewSchemaValidator
+type schemaArgs struct {
+	Schema       *spec.Schema
+	Root         interface{}
+	Path         string
+	KnownFormats strfmt.Registry
+	Options      []validate.Option
+}
+
+// RatchetingSchemaValidator wraps kube-openapis SchemaValidator to provide a
+// ValidateUpdate function which allows ratcheting
+type RatchetingSchemaValidator struct {
+	schemaArgs
+}
+
+func NewRatchetingSchemaValidator(schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, options ...validate.Option) *RatchetingSchemaValidator {
+	return &RatchetingSchemaValidator{
+		schemaArgs: schemaArgs{
+			Schema:       schema,
+			Root:         rootSchema,
+			Path:         root,
+			KnownFormats: formats,
+			Options:      options,
+		},
+	}
+}
+
+func (r *RatchetingSchemaValidator) Validate(new interface{}) *validate.Result {
+	sv := validate.NewSchemaValidator(r.Schema, r.Root, r.Path, r.KnownFormats, r.Options...)
+	return sv.Validate(new)
+}
+
+func (r *RatchetingSchemaValidator) ValidateUpdate(old, new interface{}) *validate.Result {
+	wrapped := newRatchetingValueValidator(old, r.schemaArgs, nil, nil)
+	return wrapped.Validate(new)
+}
+
+// RatchetingValueValidator is an invocation of SchemaValidator.ValidateUpdate
+// for a specific argument for `old`.
+//
+// It follows the openapi SchemaValidator down its traversal of the new value
+// by injecting validate.Option into each recursive invocation.
+//
+// It's main job is to keep the old/new values correlated as the traversal
+// continues, and postprocess errors according to our ratcheting policy.
+type ratchetingValueValidator struct {
+	schemaArgs
+
+	// Currently correlated old value during traversal of the schema/object
+	oldValue interface{}
+
+	// Validator under whose objects this instance validates a subfield or subindex.
+	// Or nil if this validates a root object
+	parent *ratchetingValueValidator
+
+	// Path from parent to this validator or nil if this is a root validator
+	keyInParent interface{}
+
+	currentInvocation *ratchetInvocationOptions
+}
+
+// Context to track validation of a specific new value during a call to `Validate`
+type ratchetInvocationOptions struct {
+	thunk *ratchetingValueValidator
+
+	// Value being validated
+	value interface{}
+
+	// Cached comparison result of DeepEqual
+	comparisonResult *bool
+
+	// Cached map representation of a map-type list, or nil if not map-type list
+	mapList common.MapList
+
+	// Children spawned by a call to `Validate` on this object
+	children map[interface{}]*ratchetInvocationOptions
```

Comment:
doc what the key is. A string or an index (for arrays) ?

Reply:
added doc